cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/assumptions/Assumption.kt

```diff
@@ -0,0 +1,261 @@
+ * result. If it is [AssumptionStatus.Rejected], nodes and edges containing them are supposed to be
+ * considered nonexistent, or not existing with the same properties. Results with assumptions and
+ * status [AssumptionStatus.Rejected] should be considered false positives.
+ */
+enum class AssumptionStatus {
+    /**
+     * Default state: Patterns(Paths) depending on the assumption are returned, result has an
+     * undecided state, assumptions are shown.
+     */
+    Undecided,
+    /**
+     * User or Algorithm: Patterns(Paths) depending on the assumption are not returned, results have
+     * a decided state, assumptions are (not?) shown.
+     */
+    Rejected,
+    /**
+     * User or Algorithm: Patterns(Paths) depending on the assumption are returned, results have a
+     * decided state, assumptions are shown.
+     */
+    Accepted,
+    /**
+     * User or Algorithm: Patterns(Paths) depending on the assumption are returned, results have a
+     * decided state, assumptions are not shown.
+     */
+    Ignored,
+}
+
+enum class AssumptionType {
+    SyntaxAmbiguityAssumption,
+    InferenceAssumption,
+    ClosedMacroAssumption,
+    UnsupportedLanguageProblem,
+    MissingCodeProblem,
+    AmbiguityAssumption,
+    ConceptAssumption,
+    ExhaustiveEnumerationAssumption,
+    CompletenessAssumption,
+    SoundnessAssumption,
+    ControlFlowAssumption,
+    CFIntegrityAssumption,
+    NoExceptionsAssumption,
+    CFAllOrNothingExecutesAssumption,
+    TrustedConfigAssumption,
+    DataFlowAssumption,
+    ExternalDataAssumption,
+    NetworkAvailableAssumption,
+    ResourceExistsAssumption,
+    ServiceReachableAssumption,
+    AtomicExecutionAssumption,
+    TrustBoundaryAssumption,
+    DataRangeAssumption,
+    TrustedInputAssumption,
+}
+
+interface HasAssumptions {
+    val assumptionNodes: MutableList<Assumption>
+
+    /**
+     * This function adds a new assumption to the object it is called on. If the object is a node or
+     * edge. The Assumption is added as an overlaying node for presentation in the graph. If the
+     * base object it is neither node nor edge, an assumptionID has to be provided to make
+     * assumptions identifiable across CPG runs.
+     *
+     * @param assumptionType The type of assumption used to differentiate between assumptions and
+     *   group similar assumptions.
+     * @param assumptionID an ID chosen by the caller, e.g. filename or classname, with a function
+     *   name and a potential counter. The assumptionID is needed if the assumption is neither on a
+     *   node, nor on an edge.
+     * @param scope The scope that the assumption has validity for, here the scope is a node,
+     *   because the assumption is valid for every node in its ast subtree.
+     * @param message The message describing the assumption that was taken.
+     */
+    fun HasAssumptions.assume(
+        assumptionType: AssumptionType,
+        message: String,
+        assumptionID: String? = null,
```

Comment:
as disussed, we want to make sure the the node ID is deterministic in the future, so we should use it instead